src/server/game/Craft/Crafting.cpp

```diff
@@ -0,0 +1,243 @@
+
+    if (_craftingData->ItemBonusTreeID)
+        bonusTreeIds.push_back(_craftingData->ItemBonusTreeID);
+
+    uint32 craftingDifficulty = _craftingData->CraftingDifficulty;
+
+    // List all reagent "slots" for this craft, with ReagentSlotID & ReagentCount
+    for (ModifiedCraftingSpellSlotEntry const* spellSlot : spellSlots)
+    {
+        // If no ReagentCount used by spell why even bother ?
+        if (spellSlot->ReagentCount < 1)
+            continue;
+
+        ModifiedCraftingReagentSlotEntry const* reagentSlot = sModifiedCraftingReagentSlotStore.LookupEntry(spellSlot->ModifiedCraftingReagentSlotID);
+        uint32 reagentCountForSlot = 0;
+
+        auto providedReagentItr = _spell->m_targets.ModifiedCraftingReagents.find(spellSlot->Slot);
+        if (providedReagentItr != _spell->m_targets.ModifiedCraftingReagents.end())
+        {
+            for (WorldPackets::Spells::SpellCraftingReagent const& providedReagentForSlot : providedReagentItr->second)
+            {
+                if (_player->HasItemCount(providedReagentForSlot.ItemID, providedReagentForSlot.Quantity))
+                {
+                    reagentCountForSlot += providedReagentForSlot.Quantity;
+                    usedReagentsToDestroy[providedReagentForSlot.ItemID] = providedReagentForSlot.Quantity;
+
+                    if (CraftingReagentQualityEntry const* reagentQuality = sDB2Manager.GetCraftingReagentQualityByItem(providedReagentForSlot.ItemID))
+                        craftingDifficulty += reagentQuality->MaxDifficultyAdjustment;
+                }
+            }
+        }
+
+        if (reagentCountForSlot < uint32(spellSlot->ReagentCount))
+        {
+            // If we don't have enough Reagent for this slot and it is required, craft failed
+            if (reagentSlot->ReagentType == MCR_REAGENT_TYPE_REQUIRED)
+                return SpellCastResult::SPELL_FAILED_NEED_MORE_ITEMS;
+
+            // Still not enough but the reagent is optional, just skip
+            continue;
+        }
+
+        std::vector<ModifiedCraftingCategoryEntry const*> categories = sDB2Manager.GetMCRCategoryByReagentSlot(spellSlot->ModifiedCraftingReagentSlotID);
+        bool hasOneMatchingItem = false;
+
+        for (ModifiedCraftingCategoryEntry const* category : categories)
+        {
+            std::vector<ModifiedCraftingReagentItemEntry const*> reagentItems = sDB2Manager.GetMCRReagentItemByCategory(category->ID);
+
+            for (ModifiedCraftingReagentItemEntry const* reagentItem : reagentItems)
+            {
+                std::set<ItemSparseEntry const*> requiredItems = sDB2Manager.GetItemSparseByMCRReagentItem(reagentItem->ID);
+
+                for (WorldPackets::Spells::SpellCraftingReagent const& providedReagent : providedReagentItr->second)
+                {
+                    bool hasItem = std::count_if(std::cbegin(requiredItems), std::cend(requiredItems), [providedReagent](ItemSparseEntry const* itemSparse) {
```

Comment:
pretty inefficient, it's best to do .contains(providedReagent.ItemID) for 1(O) lookup if you decide to pre-process them